src/messages/MOSDOp.h

```diff
@@ -48,8 +48,10 @@ class MOSDOp : public Message {
   pg_t pgid;
   bufferlist::iterator p;
   // Decoding flags. Decoding is only needed for messages catched by pipe reader.
-  bool partial_decode_needed;
-  bool final_decode_needed;
+  // Transition from true -> false without locks being held
+  // Can never see final_decode_needed == false and partial_decode_needed == true
+  volatile bool partial_decode_needed;
+  volatile bool final_decode_needed;
```

Comment:
I think that'll get expensive; don't we dereference those bools every time we look at an MOSDOp? I *think* we're just using those as flags to determine if we need to grab a lock and do the decode before looking at memory, or if we can look at the memory directly, which means presumably with careful coding it's okay if we read a stale value (we'd just grab the lock needlessly). But apparently it's being done wrongly, or maybe it really is just that the memory reads are being reordered incorrectly (which is what ```volatile``` prevents)...